dubesar/Ultimate-Java-Resources
# Issue #580: Changes BubbleSort Algorithm to generic

## What does this PR do?

Deletes previous [bubble_sort.java](https://github.com/dubesar/Ultimate-Java-Resources/blob/master/Sorting/bubble_sort.java) and adds new algorithm that sorts all generic Datatypes. 

## Test Plan

This a basic test conducted by me.

```
PS H:\GitHub\Ultimate-Java-Resources\Sorting> cat .\BubbleSort.java
package Sorting;

import java.util.Arrays;

/***************************************************
 * Sorts a array through BubbleSort.
 *
 * @author https://github.com/AkMo3
 *
 **************************************************/
public class InsertionSort<E extends Comparable<E>> {

  public static void main(String[] args) {
    String[] a = {"Sachin", "Virat", "Bhuvi", "Jadeja", "Rohit", "Ashwin"};
    sort(a);
    System.out.println(Arrays.toString(a));
  }

  /**
   * Main method to sort the required array.
   *
   * @param a - array that is to be sorted.
   */
  public static void sort(Comparable[] a) {
    for (int i = 0; i < a.length - 1; i++) {
      for (int j = 0; j < a.length - 1 - i; j++) {
        if (greater(a, j)) {
          swap(a, j);
        }
      }
    }
  }

  /**
   * Method to check whether element greater than the next.
   *
   * @param array - The array that is to be sorted.
   * @param index - index of position that is to be compared.
   */
  private static boolean greater(Comparable[] array, int index) {
    return array[index].compareTo(array[index + 1]) > 0;
  }

  /**
   * Method to swap positions of array.
   *
   * @param array - array in which array that is to be sorted
   * @param index - index that is to be swaped
   */
  private static void swap(Comparable[] array, int index) {
    Comparable temp = array[index];
    array[index] = array[index + 1];
    array[index + 1] = temp;
  }
}
PS H:\GitHub\Ultimate-Java-Resources\Sorting> java .\BubbleSort.java
Note: .\BubbleSort.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
[Ashwin, Bhuvi, Jadeja, Rohit, Sachin, Virat]
```
Steps to test-

1.  Create a new Generic Array in the main class.
2.  Import this class in your code.
3.  Call "sort" function from the above class and sort the array.
4.  Check if the array if sorted or not.

## Related PRs and Issues

Fixes: #580 

### Have you read the [Contributing Guidelines](https://github.com/dubesar/Java-A-Z/blob/master/CONTRIBUTING.md)?

Yes
